433 add save feature for search results

gracemyz opened the pull request:
* Add an "export" button to individual and summary results tabs. 
  * Keyboard shortcut is Alt-E.
  * Allowed export formats are XML, JSON, txt, and TSV. Default is XML
 * Remember recent folders for search results and search targets.

kchall commented:
Thanks, @gracemyz ! This basically looks great. A couple of cosmetic things:
1. Could we change the name of the "Export" function to "Export current results" for clarity?
2. Could we add the Export option to the "File" menu ("Export current results") so that we can see what the shortcut is?  (Or is this tricky because it's only an option when the results are displayed and not when the focus is on the search target window?)
3. If we can add it to the file menu, can we then change "Save as" to "Save target file as"?
4. The keyboard shortcut doesn't seem to be working for me on a Mac. 
5. Re: The default file name for the export: I like that it includes the "individual results" or "summary results" automatically; when I was testing it, the default file name was "None_individual_results" etc. -- I think this is because I hadn't saved the search target file, and it's great that if I had been working from a saved file, it uses that file name. Could we change the default from "None" to "SLPAA" if there is no saved name though? so e.g. "SLPAA_individual_results" etc.
6. Is there a particular reason to default to .xml? I think I would have guessed that the default would be .txt or .tsv, which more closely matches the displayed results view. And my guess is that a less-savvy user will not think to check the file extension choices and also be more likely to want .txt / not know what to do with .xml. 
7. This might not be something under our control, but I'm wondering if we can label the dropdown menu in the 'export' window with "File format" to make it super obvious that this is something that the user can actively select. For example:

Here I'm 'saving' from Excel, and the drop-down menu is labelled as "File format":
<img width="988" alt="image" src="https://github.com/user-attachments/assets/41acb21c-1d36-4691-84cb-03a0ac8c4e0c" />

But here when I'm 'saving' from SLP-AA, the drop down menu isn't labelled, so it's less obvious that I can select a file format:
<img width="615" alt="image" src="https://github.com/user-attachments/assets/6d4b4d46-a06c-40c8-8115-fb2044ec73bd" />

gracemyz commented:
@kchall: Updates!

1. Changed name of "Export" function to "Export current results".
2. 
    > Could we add the Export option to the "File" menu ("Export current results") so that we can see what the shortcut is?  
    > (Or is this tricky because it's only an option when the results are displayed and not when the focus is on the search target window?)
   
   I've added tooltips on the existing Export buttons so that the shortcut is shown on hover. Is this okay or do we still want the option in the search target window "File" menu? The focus might not be a problem, but I think we'd have to create separate options and shortcuts for individual and summary results.
3. Tried to fix the keyboard shortcut for Macs: Command-Shift-S. Does it work now? Also, does the tooltip work?
4. Changed default file name to include "SLPAA" instead of "None" if the search target file hasn't been saved.
5. > Is there a particular reason to default to .xml? I think I would have guessed that the default would be .txt or .tsv, which more closely matches the displayed results view.

     I was just thinking that a hierarchical format would be easier to interpret for results files that include multiple search targets or corpora. But we also don't have the "append to existing results file" functionality yet, so maybe results tables don't get complex enough to warrant the nice hierarchy. :) I've switched the default format to .tsv. 
6. > I'm wondering if we can label the dropdown menu in the 'export' window with "File format" to make it super obvious that this is something that the user can actively select.

   In Windows, the dropdown is already labelled with "Save as type". It _seems_ that we should be able to customize the dropdown menu label, since the [setLabelText()](https://doc.qt.io/qt-6/qfiledialog.html#DialogLabel-enum) function exists, but it doesn't do anything in the native Windows dialog. I've tried customizing it anyway - does it work for the native Mac dialog? We can also choose not to use the native dialog, but the widget-based one provided by Qt.

kchall commented:
Thanks, @gracemyz ! I was just trying to test this, but got an error message when trying to do the search:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/idlelib/run.py", line 156, in main
    ret = method(*args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/idlelib/run.py", line 559, in runcode
    exec(code, self.locals)
  File "/Users/KCH/Desktop/SLPAA/src/main/python/main.py", line 8, in <module>
    exit_code = appctxt.run()
  File "/Users/KCH/Desktop/SLPAA/src/main/python/gui/app.py", line 23, in run
    return self.app.exec_()
  File "/Users/KCH/Desktop/SLPAA/src/main/python/search/search_builder.py", line 1087, in on_search_button_clicked
    self.search_clicked.emit(button.objectName())
  File "/Users/KCH/Desktop/SLPAA/src/main/python/search/search_builder.py", line 223, in handle_search_clicked
    self.results_view = ResultsView(resultsdict, mainwindow=self)
  File "/Users/KCH/Desktop/SLPAA/src/main/python/search/results.py", line 65, in __init__
    summarylayout.addWidget(self.create_toolbar("summary"))
  File "/Users/KCH/Desktop/SLPAA/src/main/python/search/results.py", line 103, in create_toolbar
    action_export.setToolTip(QKeySequence.listToString(export_shortcuts[export_num], QKeySequence.NativeText))
TypeError: listToString(Iterable[Union[QKeySequence, QKeySequence.StandardKey, str, int]], format: QKeySequence.SequenceFormat = QKeySequence.PortableText): argument 1 has unexpected type 'StandardKey'



The specific search I tried was on the sample corpus, looking for either a B2 handshape or a movement module with hooking, and doing either of those searches by itself also triggers the same error.

gracemyz commented:
@kchall I hope this does the trick... could you let me know if it works on your Mac now?
